### PR TITLE
[MSSQL] Implement MS_Description support for procedures, functions and parameters

### LINKED ISSUE #2256: SQL Server's MS_Description is ignored on stored procedures and functions
Linq2db seems to ignore descriptions attached to stored procedures, functions and their parameters. I would expect it to generate /// comments. On the contrary MS_description extended properties are respected on at least table columns.

Is it technically possible to extend SqlServerSchemaProvider.cs and LinqToDB.SqlServer.ttinclude in such as way that MS_description will be put into /// comments?

### PR DESCRIPTION
**Pull request by MaceWindu**
Fix #2256

### CODE REVIEW

**Review by sdanyliv — APPROVED** [review]

File: Source/LinqToDB.Templates/DataModel.ttinclude [hunk 1]
```diff
@@ -318,7 +319,8 @@ void LoadServerMetadata(DataConnection dataConnection)
 						ParameterType = pr.ParameterType,
 						SystemType    = pr.SystemType,
 						DataType      = pr.DataType.ToString(),
-						IsNullable    = pr.IsNullable
+						IsNullable    = pr.IsNullable,
+						Description   = pr.Description
```
Comment (viceroypenguin) [inline Source/LinqToDB.Templates/DataModel.ttinclude:323]:
```suggestion
						Description   = pr.Description,
```

adding a `,` to the end of the last item means it's not a change on future adds to the list.

**Review by viceroypenguin — APPROVED** [review]

File: Tests/Linq/DataProvider/SqlServerTests.cs [hunk 2]
```diff
@@ -1817,5 +1817,40 @@ public void Issue1468Test([IncludeDataSources(false, TestProvName.AllSqlServer)]
 				Assert.IsNull(proc.ResultException);
 			}
 		}
+
+		[Test]
+		public void TestDescriptions([IncludeDataSources(false, TestProvName.AllSqlServer)] string context)
```
Comment (Shane32) [inline Tests/Linq/DataProvider/SqlServerTests.cs:1822]:
```suggestion
		public void TestDescriptions([IncludeDataSources(false, TestProvName.AllSqlServer2005Plus)] string context)
```

**Review by Shane32 — CHANGES_REQUESTED** [review]
Test runs fine on my PC with 2005, 2008, 2012, 2014, and 2017, but not SQL Server 2000.

File: Source/LinqToDB.Templates/DataModel.ttinclude (hunk 1 above)
Comment (Shane32) [reply Source/LinqToDB.Templates/DataModel.ttinclude:323]: :+1: If there was a format rule to enforce that for multi-line syntax, it wouldn't be a bad idea.

**Comment by MaceWindu** [comment]
@Shane32 could you retest sql2000?

**Review by viceroypenguin — APPROVED** [review]

**Comment by Shane32** [comment]
Sure -- same error: `Invalid object name: 'SYS.EXTENDED_PROPERTIES'.`

```
System.Data.SqlClient.SqlException : Invalid object name 'SYS.EXTENDED_PROPERTIES'.
Data:
  HelpLink.ProdName: Microsoft SQL Server
  HelpLink.ProdVer: 08.00.2039
  HelpLink.EvtSrc: MSSQLServer
  HelpLink.EvtID: 208
  HelpLink.BaseHelpUrl: http://go.microsoft.com/fwlink
  HelpLink.LinkId: 20476

DataConnection: BeforeExecute
-- SqlServer.2000

SELECT cmptlevel FROM master.dbo.sysdatabases WHERE name = db_name()

DataConnection: Query Execution Time (AfterExecute): 00:00:00.0807656

DataConnection: BeforeExecute
-- SqlServer.2000


				SELECT
					SPECIFIC_CATALOG + '.' + SPECIFIC_SCHEMA + '.' + SPECIFIC_NAME                as ProcedureID,
					SPECIFIC_CATALOG                                                              as CatalogName,
					SPECIFIC_SCHEMA                                                               as SchemaName,
					SPECIFIC_NAME                                                                 as ProcedureName,
					CASE WHEN ROUTINE_TYPE = 'FUNCTION'                         THEN 1 ELSE 0 END as IsFunction,
					CASE WHEN ROUTINE_TYPE = 'FUNCTION' AND DATA_TYPE = 'TABLE' THEN 1 ELSE 0 END as IsTableFunction,
					CASE WHEN SPECIFIC_SCHEMA = 'dbo'                           THEN 1 ELSE 0 END as IsDefaultSchema,
					ISNULL(CONVERT(varchar(8000), x.Value), '')                                   as Description
				FROM
					INFORMATION_SCHEMA.ROUTINES
					LEFT JOIN SYS.EXTENDED_PROPERTIES x
						ON OBJECT_ID('[' + SPECIFIC_SCHEMA + '].[' + SPECIFIC_NAME + ']') = x.major_id AND
							x.name = 'MS_Description' AND x.class = 1

DataConnection: Error
Exception: System.Data.SqlClient.SqlException
Message  : Invalid object name 'SYS.EXTENDED_PROPERTIES'.
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlDataReader.TryConsumeMetaData()
   at System.Data.SqlClient.SqlDataReader.get_MetaData()
   at System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString, Boolean isInternal, Boolean forDescribeParameterEncryption, Boolean shouldCacheForAlwaysEncrypted)
   at System.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean async, Int32 timeout, Task& task, Boolean asyncWrite, Boolean inRetry, SqlDataReader ds, Boolean describeParameterEncryptionRequest)
   at System.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, String method, TaskCompletionSource`1 completion, Int32 timeout, Task& task, Boolean& usedCache, Boolean asyncWrite, Boolean inRetry)
   at System.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, String method)
   at System.Data.SqlClient.SqlCommand.ExecuteReader(CommandBehavior behavior, String method)
   at System.Data.SqlClient.SqlCommand.ExecuteDbDataReader(CommandBehavior behavior)
   at System.Data.Common.DbCommand.ExecuteReader(CommandBehavior behavior)
   at LinqToDB.Data.DbCommandProcessor.DbCommandProcessorExtensions.ExecuteReaderExt(IDbCommand cmd, CommandBehavior commandBehavior) in C:\Users\Shane\source\repos\linq2db\Source\LinqToDB\Data\DbCommandProcessor\DbCommandProcessorExtensions.cs:line 42
   at LinqToDB.Data.DataConnection.ExecuteReader(IDbCommand command, CommandBehavior commandBehavior) in C:\Users\Shane\source\repos\linq2db\Source\LinqToDB\Data\DataConnection.cs:line 1387
   at LinqToDB.Data.DataConnection.ExecuteReader(CommandBehavior commandBehavior) in C:\Users\Shane\source\repos\linq2db\Source\LinqToDB\Data\DataConnection.cs:line 1419
```

**Comment by MaceWindu** [comment]
I see, will remove support from 2000 version

**Comment by Shane32** [comment]
Just FYI, there's been a bunch of other tests added since I had cleaned up all the SQL 2000 tests that don't work either.
* ExecProcedureAsyncTestAnonymParam
* ExecProcedureTestAnonymParam
* ExecProcedureTestAnonymParamAsync
* ExecProcedureTestAnonymParamGeneric
* InOutProcedureTest
* InOutProcedureTestAsync
* TestIssue1144

Those are just under Tests.DataProvider.SqlServerTests; I didn't run a full test.  I can run a full set of tests if you like.  Or I can add a PR that changes `AllSqlServer` to `SqlServer2005Plus` for all those tests.

**Comment by MaceWindu** [comment]
I will mark them as 2005+, as we didn't added required objects to create db script for 2000

**Comment by MaceWindu** [comment]
@Shane32 , why those tests fail? I don't see anything wrong with them:
ExecProcedureAsyncTestAnonymParam
ExecProcedureTestAnonymParam
ExecProcedureTestAnonymParamAsync
ExecProcedureTestAnonymParamGeneric

**Comment by Shane32** [comment]
I'm running the tests again.  To begin with, `CreateDatabase` is failing due to: `Could not find stored procedure 'sys.sp_addextendedproperty'.`  In order to properly run the rest of the tests, we would probably need to fix that.

**Comment by Shane32** [comment]
So I probably have an old copy of the SQL 2000 create script

**Comment by MaceWindu** [comment]
I've removed sp_addextendedproperty for sql2000 as we don't test it anyway, but it is strange - this procedure should be in v2000

**Comment by Shane32** [comment]
Let me play with it for a minute.....

**Comment by Shane32** [comment]
Ok, so removing the last bit of the create script (where it references `sys.sp_addextendedproperty` fixes the issue with the tests.

These now run fine:
* ExecProcedureAsyncTestAnonymParam
* ExecProcedureTestAnonymParam
* ExecProcedureTestAnonymParamAsync
* ExecProcedureTestAnonymParamGeneric
* InOutProcedureTest
* InOutProcedureTestAsync

Still fails:
* TestIssue1144

I'll re-run the rest of the tests also.

**Comment by MaceWindu** [comment]
I've disabled TestIssue1144 for sql2000, as it tests issue with description anyway

**Comment by Shane32** [comment]
> I've removed sp_addextendedproperty for sql2000 as we don't test it anyway, but it is strange - this procedure should be in v2000

It is `dbo.` not `sys.` in sql2000 and has different arguments, I believe.  It did give this error:

```
System.Data.SqlClient.SqlException : Could not find stored procedure 'sys.sp_addextendedproperty'.
```

After changing it to `dbo.` I get this error:

```
System.Data.SqlClient.SqlException : An invalid parameter or option was specified for procedure 'sp_addextendedproperty'.
```

**Comment by MaceWindu** [comment]
I see. Anyway, it doesn't make sense to implement support for v2000 for now as it will slowdown schema retrieval greatly and probably nobody use it anymore :)

**Comment by Shane32** [comment]
I agree.

**Comment by Shane32** [comment]
And it doesn't work at all with .Net Core anyway.

**Comment by Shane32** [comment]
Just FYI, the rest of the test run completed with these failures for SQL 2000:
* Issue1198Test - only the LinqService failed - `LinqToDB.LinqToDBException : QuerySerializer error. Unknown object 'LinqToDB.SqlQuery.SelectQuery'.`
* ValueConversionTests.BoolJoinTest - `The text, ntext, and image data types cannot be compared or sorted, except when using IS NULL or LIKE operator.`
* ValueConversionTests.BoolNotTest - `The text, ntext, and image data types cannot be compared or sorted, except when using IS NULL or LIKE operator.`
* ValueConversionTests.BoolTest - `The text, ntext, and image data types cannot be compared or sorted, except when using IS NULL or LIKE operator.`
* SchemaProviderTests.PrimaryForeignKeyTest - `Invalid column name 't1'.`
* SchemaProviderTests.Test - `Invalid column name 't1'.`
* Issue681Tests.TestTableFQN - `Invalid column name 't1'.`
* Issue792Tests.TestWithTransactionThrowsFromProvider - `Expected: instance of <System.Exception> But was:  null`
* Issue792Tests.TestWithoutTransaction - `Invalid column name 't1'.`
* DropTableTests.DropSpecificDatabaseTableTest - `Invalid column name 't1'.`

**Pull request merged**